src/Pact/Native/Capabilities.hs

```diff
@@ -326,3 +334,51 @@ emitEventDef =
         DefcapManaged {} -> return ()
         DefcapEvent -> return ()
       _ -> evalError' i $ "emit-event: must be managed or event defcap"
+
+createPrincipalDef :: NativeDef
+createPrincipalDef =
+  defRNative "create-principal" createPrincipal'
+  (funType tTyString
+    [ ("guard", tTyGuard Nothing) ] )
+  [LitExample "(create-principal (read-keyset 'keyset))"]
+  "Create a principal which unambiguously identifies GUARD."
+  where
+    createPrincipal' :: RNativeFun e
+    createPrincipal' _i [TGuard g _] =
+      pure $! toTerm $ createPrincipal g
+    createPrincipal' i as = argsError i as
+
+createPrincipal :: Guard (Term Name) -> Text
+createPrincipal = \case
+  GPact (PactGuard pid n) ->
+    "p:" <> asString pid <> ":" <> n
+  GKeySet (KeySet ks pf) -> case (toList ks,asString pf) of
+    ([k],"keys-all") ->
+      "k:" <> asString k
+    (l,fun) ->
+      "w:" <> toHash (map _pubKey l) <> ":" <> fun
+  GKeySetRef (KeySetName n) ->
+    "r:" <> n
+  GModule (ModuleGuard mn n) ->
+    "m:" <> asString mn <> ":" <> n
+  GUser (UserGuard uf args) ->
+    "u:" <> asString uf <> ":" <> toHash (map toJSONPactValue args)
```

Comment:
having second thoughts about involving aeson here. Maybe it's finally time for a `Serialize` instance for `PactValue`

Reply:
A counterargument is that PactValue JSON is all over our blockchain :\

Reply:
I'd be fine with moving to `Serialize` for efficiency and accuracy's sake. Fewer things reprsented as `aeson` types the better imo. However, the surface area for messing that up is massive. I'd want to make sure we have full attention on that if we do it.

Reply:
This deserves a followup PR with a `to/fromEncoding` migration before `Serialize`, as discussed.